doc/specs/#492 - Default Terminal/spec.md

```diff
@@ -0,0 +1,248 @@
+    - `HRESULT ConsoleEstablishHandoff(HANDLE server, HANDLE driverInputEvent, const PortableArguments* const args, const PortableConnectMessage* const msg)`
+        - `HANDLE server`: This is the server side handle to the console driver, used with `DeviceIoControl` to receive/send messages with the client command-line application
+        - `HANDLE driverInputEvent`: The input event is created and assigned to the driver immediately on first connection, before any messages are read from the driver, to ensure that it can track a blocking state should first message be an input request that we do not yet have data to fill. As such, the inbox console will have created this and assigned it to the driver before pulling off the connection packet and determining that it wants to delegate. Therefore, we will transfer ownership of this event to the updated console.
+        - `const PortableArguments* const args`: This contains the startup argument information that was passed in when the process was started including the original command line and the in/out handles.
+            - The `ConsoleArguments` structure could technically change between versions, so we will make a version agnostic portable structure that just carries the communication from the old one to the new one.
+        - `const PortableConnectMessage* const msg`:
+            - The `CONSOLE_API_MSG` structure contains both the actual packet data from the driver as well as some overhead/administration data related to the packet state, ordering, completion, errors, and buffers. It's a broad scope structure for every type of message we process and it can change over time as we improve the way the `conserver.lib` handles packets.
+            - This represents a version agnostic variant for ONLY the connect message that can pass along the initial connect information structure, the packet sequencing information, and other relevant payload only to that one message type. It will purposefully discard references to things like a specific set of API servicing routines because the point of handing off is to get updated routines, if necessary.
+            
+        - *Return* `HRESULT`: This is one of the older style methods in the initialization. We moved them from mostly `NTSTATUS` to mostly `HRESULT` a while ago to take advantage of `wil`. This one will continue to follow the pattern and not move to exceptions. A return of `S_OK` will symbolize that the handoff worked and the inbox console can clean itself up and stop handling the session.
+- When the connection packet is parsed for visibility information (see `srvinit.cpp`), we will attempt to resolve the registered handoff and call it.
+    - In the initial revision here, I have this as a `LoadLibrary`/`GetProcAddress` to the above exported contract method from the updated console. This maintains the server session in the same process space and avoids:

```

Comment:
HYPOTHETICALLY, the user could configure the DelegationConsole to some other random binary with that same function name and signature, right? It doesn't _need_ to be an openconsole. 

So if Hyper wanted so install a specific conhost version, they _could_. Or they could get crazier and install `myConsoleServer.exe` as the delegated console instead, right?

Reply:
Is it possible to add an entry in the "Default Apps" system setting page? 

![image](https://user-images.githubusercontent.com/4710575/98320882-cae44f80-201e-11eb-92d2-a56d41000438.png)

I don't know how it works internally. But I think that there's must be some kind of protocol that allows an app registered as a "mail app", or "browser app" for this to work, right? So terminals like WT and Hyper could register themselves as "terminal app" to make an entry here, preventing `some other random binary` being usable as an alternative terminal

Reply:
We're looking into various options in the Settings panel including this page, the developer settings page, the apps and features pages, and protocol handling pages. The end goal will present a list/drop down of registered terminals or consoles, not just random binaries, on whatever page it lands on.

Reply:
We put it on the developer settings page .